### PR TITLE
AsmJit update

### PR DESCRIPTION
**Pull request by kobalicek**
I updated asmjit from oldstable to master (known as next-wip during development).

I only tested such changes on Linux so it definitely needs Windows testing. I haven't found any problems during playing, let me know if you have a list of challenging wads so I can do more tests myself.

Problems:

  [x] Asmjit now requires C++11 so all tests where the compiler is not invoked with proper flags fail. I think there should be another PR that would change the build system to use C++11 by default. All compilers setup by CI have C++11 support.

  [ ] In `jit_runtime.cpp` both functions `CreateUnwindInfoWindows()` and `CreateUnwindInfoUnix()` duplicate something that asmjit already does. It would be nice to improve asmjit to provide this information instead of calculating it on gzdoom side as this is considered internal and an internal change in asmjit would silently break gzdoom in a way that could be really difficult to debug.

### CODE REVIEW

**Comment by dpjudas** [comment]
Ah, very nice! Thanks for upgrading it to the latest version. And thanks for writing asmjit itself! I'll give it a spin on Windows this weekend.

About jit_runtime.cpp, I think almost everything in that file should ideally become part of asmjit itself. As the author of that file, you have my permission to take anything from that file and relicense it under the asmjit license. You may already know this, but just in case you don't, the unwind information is important to register as it allows throws from functions called by asmjit, allows debuggers to display the call stack (*), and enables stuff like structured exception handling to work.

I originally wanted to submit it as a PR for asmjit, but never got around to it. Partly because next-wip at the time made it seem like I'd be submitting it to the old version anyway. :)

*) This is currently only the case on Windows. But mostly because gdb doesn't tap into the information already given to the unwinder. I'm sure that one day someone will improve their debugger to also take advantage of it.

**Comment by kobalicek** [comment]
@dpjudas I will check it out although I cannot guarantee that I would progress with the unwinding at the moment as I have too many things to do. I would first try to fix asmjit so it exposes the information to build the tables, that would be a nice first step I think.

@alexey-lysiuk I can fix that, since the PR is so huge I wasn't sure whether to include the fix here or whether to make another PR that would fix the build first and make it clear that gzdoom requires C++11. But according to my testing `-std=c++11` is not enough as `rendering/hwrenderer` won't compile as it requires generic lambdas (lambda with `auto`), which were introduced in C++14.

I have found some other possible problems:
  - If you want to support clang-cl on Windows you have to use `if ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "MSVC" OR "x${CMAKE_CXX_SIMULATE_ID}" STREQUAL "xMSVC")` instead of `if (MSVC)`
  - The proper check for GCC/Clang is `if ("${CMAKE_CXX_COMPILER_ID}" MATCHES "^(GNU|Clang|AppleClang)$")`, at least the apple bundled `Clang` haven't worked for me on recent OSX.
  - To set the required C++ standard there is a cmake way to do it per target: `target_compile_features(target PUBLIC cxx_std_11)`, which would automatically propagate the setting to all targets that use that target as dependency, this could possible fix the C++14 usage as well, but...

In general I didn't want to make much changes to the build system as I'm unable to test all combinations, the conclusion is that C++11 doesn't work on my machine, I need to build gzdoom with C++14 otherwise I would get errors like:

```
../src/rendering/hwrenderer/textures/hw_texcontainer.h:56:48: error: 'auto' not allowed in lambda parameter
                unsigned index = hwTex_Translated.FindEx([=](auto &element)
                                                             ^~~~
```

**Comment by alexey-lysiuk** [comment]
The change I linked affects AsmJit library only. C++14 is already enabled for code that requires it, i.e. for `src` directory.

Personally, I don't care about non-default compilers, like Clang on Windows, MinGW, etc.
The exception is Clang on Linux because GCC code generation with `-O3` is full of surprises.
If anyone wants to support such compilers, he/she should provide PRs for that.

**Comment by kobalicek** [comment]
@alexey-lysiuk Thanks that makes sense, I completely missed that you do enable C++14 for `src`, I will update it.

File: libraries/asmjit/CMakeLists.txt [hunk 1]
```diff
@@ -1,17 +1,21 @@
 cmake_minimum_required(VERSION 2.8.7)
 
-#make_release_only()
```
Comment (alexey-lysiuk) [inline libraries/asmjit/CMakeLists.txt]: Could you please bring `make_release_only()` back?

File: src/scripting/vm/jit.cpp [hunk 2]
```diff
@@ -23,7 +23,11 @@ JitFuncPtr JitCompile(VMScriptFunction *sfunc)
 		CodeHolder code;
 		code.init(GetHostCodeInfo());
 		code.setErrorHandler(&errorHandler);
-		code.setLogger(&logger);
+
+    //FileLogger fileLogger(stdout);
+    //fileLogger.addFlags(FormatOptions::kFlagAnnotations | FormatOptions::kFlagDebugRA | FormatOptions::kFlagDebugPasses);
+    //code.setLogger(&fileLogger);
+    code.setLogger(&logger);
```
Comment (alexey-lysiuk) [inline src/scripting/vm/jit.cpp]: Please revert this "change".

**Comment by kobalicek** [comment]
I have updated the PR accordingly to the review. I'm not sure whether it would make sense to merge it immediately, I would still like more people to test it with their favorite wads to make sure it works correctly.

In addition, I will check out what I can done on asmjit side to make the unwinding code simpler.

**Comment by dpjudas** [comment]
I merged this PR into an asmjitupdate branch for further exposure and testing. It works fine on Windows with what I tested with. This PR got the green signal from me - but I'll let it be up to Graf to decide if it has gotten enough testing to be merged into master.

About the unwind code, the tricky part is to decide where to put it. The emitProlog function generates the information needed, but future more setups could be more complex. In a normal executable this data is stored in the .pdata/.eh_frame section, so one could argue CodeHolder should simply get a new Section for it. The compiler would then generate it roughly at the same point it calls emitProlog.

**Comment by dpjudas** [comment]
Looking slightly closer at the PE File Format specs I can see the .pdata section only contains the function table and then it uses offsets into .rdata where the actual unwindinfo is then stored.

The question here is how closely asmjit's sections should map to the platform specific sections?

If the goal is to have a 1:1 mapping, then the win32 version needs to store the output of `CreateUnwindInfoWindows()` in a .rdata section, and then generate an additional .pdata section that is effectively a memory dump of the RUNTIME_FUNCTION struct (one for each FuncNode).

The Linux/Mac version is somewhat simpler. It only needs the output of `CreateUnwindInfoUnix` stored in an .eh_frame section.

**Comment by dpjudas** [comment]
Looking slightly closer at the PE File Format specs I can see the .pdata section only contains the function table and then it uses offsets into .rdata where the actual unwindinfo is then stored.

The question here is how closely asmjit's sections should map to the platform specific sections?

If the goal is to have a 1:1 mapping, then the win32 version needs to store the output of `CreateUnwindInfoWindows()` in a .rdata section, and then generate an additional .pdata section that is effectively a memory dump of the RUNTIME_FUNCTION struct (one for each FuncNode).

The Linux/Mac version is somewhat simpler. It only needs the output of `CreateUnwindInfoUnix` stored in an .eh_frame section.

**Comment by dpjudas** [comment]
I tried to change the implementation to create .rdata and .pdata sections instead (with the assumption that code would then be moved internally to asmjit).

Unfortunately adding any section, even an empty one, causes the code generation to somehow fail. I'm not sure if `CodeHolder::newSection` has a bug or if I'm doing something wrong.

This is the commit illustrating it: https://github.com/coelckers/gzdoom/commit/1dfa6702783f5ff9933eeae0557a6891eddc2557
It can be found on the asmjitsection branch in the gzdoom repository.

**Comment by madame-rachelle** [comment]
I'll throw this right into QZDoom so that it can get some public testing. @kobalicek do you have a forum account? If there are any issues they'll appear here: https://forum.zdoom.org/viewforum.php?f=2 - A forum account is not required, but if you wish to interact with the people who will be testing this, it certainly will make things a lot easier for you.

**Comment by kobalicek** [comment]
@dpjudas I see, sections are a relatively new feature in asmjit so it's possible there is a bug or the API is misused; I can definitely check out the branch.

@madame-rachelle Thanks for more testing. I believe it would be okay, but if there is a discussion I would like to participate I would create the account.

**Comment by nashmuhandes** [comment]
@kobalicek hi, at the advice of @dpjudas I would like to report a crashing bug with the asmjit update. Is there any way I can deliver a private message to you? I would need to send you a WIP copy of a medium-sized project we are working on, because I am unable to reproduce the bug if I isolated it to a minimal example.

image

The other reason I would prefer to send it to you privately is because I cannot share the project publicly at this time. :)

Please advise. Thanks!

**Comment by kobalicek** [comment]
Hi @nashmuhandes, you can contact me at kobalicek.petr [gmail.com], but please be absolutely sure that the bug is caused by the asmjit update. I can definitely take a look. So far nobody complained about the register allocator so either you are the first or the bug is not RA related.

**Comment by nashmuhandes** [comment]
> Hi @nashmuhandes, you can contact me at

Done. Thanks!

**Comment by kobalicek** [comment]
Good news is that I can repro the crash, bad news is that at the moment I have no idea where I should start as the log file that I generated has 180k lines. I will try investigating the problem.

This is actually what I expected from this PR - to identify potential issues with the new version.

**Comment by kobalicek** [comment]
So I have identified the problem, I don't have a fix yet, but I know where to look.

asmjit issue:
  - https://github.com/asmjit/asmjit/issues/251

**Comment by nashmuhandes** [comment]
@kobalicek glad that you've found something :) Thanks for looking into this!

**Comment by kobalicek** [comment]
I have found the issue and it seems it's not related to asmjit. GZDoom generates the following instruction, that would pass without enabling strict validation feature:

```x86asm
mov tmpPtr1, regD7
```

Where tmpPtr1 is 64-bit register and `regD7` is 32-bit register. In previous version AsmJit just promoted the second operand to 64-bit so normally nothing would happen, however, the new version can use memory load instead of using the register, which would turn it to:

```x86asm
mov rax, dword [rsp+XXX]
```

Which would be also promoted without strict validation to:

```x86asm
mov rax, qword [rsp+XXX]
```

And now you can see the problem.

It's easy to fix, but I have to find the place where this happens first.

**Comment by kobalicek** [comment]
Updated - The crash should be eliminated now and all the generated assembly by GZDoom should be conforming to asmjit's strict validation feature (I have tested it, but not enabled it by default as it comes with a cost). I had to change few things, most changes were about properly handling some constructs like `setxxx` that expect 8-bit regs, addresses `[a + b]` that expect 64-bit regs, etc... 

More testing welcome.

@nashmuhandes Would love to play that was once it's finished :)

**Comment by nashmuhandes** [comment]
@kobalicek Wonderful! I just tested your engine fixes with the unstripped version of the project, as well as another much larger project, and both of them run flawlessly. :) Thank you for the fix, and also for expressing interest in my project! I will definitely inform you when it has made more progress.

As far as further testing, so far, playing some other GZDoom mods also seems to be without problems... hopefully I don't find any more. :P

**Comment by madame-rachelle** [comment]
So far it seems like everything has gone good with this, I think it should be safe to merge. Nothing exploded in QZDoom since the most recent update to this PR.

**Comment by coelckers** [comment]
Ok, but how does this relate to the work branch? I don't want to merge this into master and then create conflicts with dpJudas's work over there.

**Comment by dpjudas** [comment]
You can safely merge this without worrying about those branches. They were for testing anyway.

**Comment by kobalicek** [comment]
I have noticed this branch is already behind almost 100 commits, I can rebase it.

**Comment by madame-rachelle** [comment]
This had to be reverted because it breaks GZDoom's exception handling, and with how it handles the JITted code (it often comes from the user before it is compiled), this is critical for proper functionality.

We have a forum where we are discussing this - if you create an account I can give you access to that discussion, but I'll need to know your username. https://forum.zdoom.org/

**Pull request merged**